L1Trigger/Phase2L1GMT/src/TPSAlgorithm.cc

```diff
@@ -193,103 +205,133 @@ propagation_t TPSAlgorithm::propagate(const ConvertedTTTrack& track, uint layer)
+    if (BITSHIFTCURVSCALEC2CORRS[layer] != 0) {
+      dphi2_c2Full += (1 << (BITSHIFTCURVSCALEC2LEADS[layer] + BITSHIFTCURVSCALEC2CORRS[layer] + 1));
+      dphi2_c2Full -=
+          ((absKpad << (BITSHIFTCURVSCALEC2CORRS[layer] + 1)) + (1 << (2 * BITSHIFTCURVSCALEC2CORRS[layer])));
+    }
+    dphi_c2 += (dphi2_c2Full >> BITSHIFTPROP2C2);
+
+    ap_uint<BITSPROPCOORD + 2 * BITSTTCURV - 2> dphi3_c2Full = (k2pad << 1) +
+                                                               (1 << (2 * BITSHIFTCURVSCALEC2LEADS[layer])) -
+                                                               (absKpad << (BITSHIFTCURVSCALEC2LEADS[layer] + 1));
+    if (BITSHIFTCURVSCALEC2CORRS[layer] != 0) {
+      dphi3_c2Full +=
+          ((1 << (2 * BITSHIFTCURVSCALEC2CORRS[layer])) + (absKpad << (BITSHIFTCURVSCALEC2CORRS[layer] + 1)));
+      dphi3_c2Full -= (1 << (BITSHIFTCURVSCALEC2LEADS[layer] + BITSHIFTCURVSCALEC2CORRS[layer] + 1));
+    }
+    dphi3_c2Full *= prop3_coord2;
+    if (layer == 0 or layer == 4) {
+      dphi_c2 += (dphi3_c2Full >> BITSHIFTPROP3C2);
+    } else {
+      dphi_c2 -= (dphi3_c2Full >> BITSHIFTPROP3C2);
+    }
+  }
 
-  if (negativeCurv == 1)
-    dphi = -dphi;
+  if (dphi_c2 > PROPMAX)
+    dphi_c2 = PROPMAX;
+  else if (dphi_c2 < -PROPMAX)
+    dphi_c2 = -PROPMAX;
+
+  //subtract the magnitude for positive k, add the magnitude for negative k
+  if (negativeCurv == 0)
+    dphi_c2 = -dphi_c2;
 
   if (is_barrel)
-    out.coord2 = -dphi / PHIDIVIDER;
+    out.coord2 = (dphi_c2 >> PHISHIFT);
   else
-    out.coord2 = (phi - dphi) / PHIDIVIDER;
+    out.coord2 = ((phi + dphi_c2) >> PHISHIFT);
 
-  ap_int<BITSETA> eta = track.eta();
-  out.eta = eta / ETADIVIDER;
+  ap_uint<BITSPROPSIGMACOORD_B + BITSTTCURV - 1> s1kFull = res1_coord1 * absK;
+  ap_uint<BITSPROPSIGMACOORD_B + BITSTTCURV - 1 - BITSHIFTRES1> s1k = res0_coord1 + (s1kFull >> BITSHIFTRES1);
+  s1k = s1k >> PHISHIFT;
+  if (s1k >= SIGMAMAX)
+    out.sigma_coord1 = SIGMAMAX;
+  else if (s1k < SIGMAMIN)
+    out.sigma_coord1 = SIGMAMIN;
+  else
+    out.sigma_coord1 = s1k;
 
-  ap_uint<2 * BITSTTCURV - 2> curvature2All = curvature * curvature;
-  ap_uint<BITSTTCURV2> curvature2 = curvature2All / 2;
-  /*
-  /////New propagation for sigma
-  ap_uint<BITSTTCURV - 1> absK = 0;
-  if (track.curvature() < 0)
-    absK = ap_uint<BITSTTCURV - 1>(-track.curvature());
+  ap_uint<BITSPROPSIGMACOORD_B + BITSTTCURV - 1> s2kFull = res1_coord2 * absK;
+  ap_uint<BITSPROPSIGMACOORD_B + BITSTTCURV - 1 - BITSHIFTRES1> s2k = res0_coord2 + (s2kFull >> BITSHIFTRES1);
+  s2k = s2k >> PHISHIFT;
+  if (s2k >= SIGMAMAX)
+    out.sigma_coord2 = SIGMAMAX;
+  else if (s2k < SIGMAMIN)
+    out.sigma_coord2 = SIGMAMIN;
   else
-    absK = ap_uint<BITSTTCURV - 1>(track.curvature());
-  */
-  //bound the resolution propagation
-  //if (absK > 6000)
-  //  absK = 6000;
+    out.sigma_coord2 = s2k;
+
+  ap_int<BITSETA> eta = track.eta();
+  out.eta = eta / ETADIVIDER;
+  //out.eta = eta >> ETASHIFT;
```

Comment:
In the future this commented out line can be removed.

Reply:
Yeah, we can remove this line in the next PR.